shell/ssh/src/main/java/org/apache/karaf/shell/ssh/Activator.java

```diff
@@ -189,6 +194,20 @@ protected SshServer createSshServer(SessionFactory sessionFactory) {
         server.setKeyExchangeFactories(SshUtils.buildKexAlgorithms(kexAlgorithms));
         server.setSignatureFactories(SshUtils.buildSigAlgorithms(sigAlgorithms));
         server.setShellFactory(new ShellFactoryImpl(sessionFactory));
+        server.addSessionListener(new SessionListener() {
+            @Override
+            public void sessionDisconnect(org.apache.sshd.common.session.Session session, int reason, String msg, String language, boolean initiator) {
+                try {
+                    Collection<ServiceReference<ChannelResourceCleaner>> references = bundleContext.getServiceReferences(ChannelResourceCleaner.class, null);
+                    for (ServiceReference<ChannelResourceCleaner> reference : references) {
+                        ChannelResourceCleaner cleaner = bundleContext.getService(reference);
+                        cleaner.close();
```

Comment:
This appears to call .close() on _every_ reference for every session that closes. Is that the correct association? 

Wouldn't we want to close only the instance of the command associated with that session?

Reply:
That's right it's a bit agressive: if we have several ssh clients on the Karaf ssh server, we will close *all* references whatever is the client.
It's pretty hard to "link" the session with the command (or maybe we can use a session property to identify the correct session).

Reply:
While we're at adding things to the Command/Action session.. this change is needed due to JDK removal of Security Manager: https://issues.apache.org/jira/browse/KARAF-7805